servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/SpliceFlatStreamToMetaSingle.java

```diff
@@ -290,26 +305,37 @@ public void onComplete() {
                             EMPTY_COMPLETED_DELIVERED)) {
                         ((PublisherSource.Subscriber<Payload>) maybeSubscriber).onComplete();
                     } else {
-                        ((PublisherSource.Subscriber<Payload>) maybeSubscriber).onError(new IllegalStateException(
-                                "Duplicate Subscribers are not allowed. Existing: " + maybeSubscriber +
-                                        ", failed the race with a duplicate, but neither has seen onNext()"));
+                        terminateWithIllegalStateException((PublisherSource.Subscriber<Payload>) maybeSubscriber);
                     }
                 } else if (!metaSeenInOnNext) {
                     ensureResultSubscriberOnSubscribe();
                     dataSubscriber.onError(new IllegalStateException(
                             "Stream unexpectedly completed without emitting any items"));
+                } else if (maybeSubscriber == EMPTY_COMPLETED_DELIVERED) {
+                    LOGGER.debug("Discarding a terminal complete from upstream because the payload publisher was " +
+                            "already terminated");
                 }
             }
         }
 
         private void ensureResultSubscriberOnSubscribe() {
-            assert !metaSeenInOnNext;
+            assert !metaSeenInOnNext : "Already seen meta-data";
             if (!onSubscribeSent) {
                 onSubscribeSent = true;
                 // Since we are going to deliver data or a terminal signal right after this,
                 // there is no need for this to be cancellable.
                 dataSubscriber.onSubscribe(IGNORE_CANCEL);
             }
         }
+
+        private void terminateWithIllegalStateException(PublisherSource.Subscriber<Payload> subscriber) {
+            subscriber.onError(new IllegalStateException("Duplicate Subscribers are not allowed. Existing: " +
+                    subscriber + ", failed the race with a duplicate, but neither has seen onNext()"));
+        }
+
+        private static StacklessCancellationException newCancellationException() {
+            return StacklessCancellationException.newInstance(
```

Comment:
is this expected to be common enough where it needs to be stackless? if not, its called from 2 different locations so we should differentiate these two cases.